src/test/java/htsjdk/samtools/cram/encoding/ReadFeaturesTest.java

```diff
@@ -25,4 +29,29 @@ public void faultyEquality() {
         final SoftClip sc2 = new SoftClip(0, new byte[] {});
         Assert.assertEquals(sc1, sc2);
     }
+
+    @Test
+    public void testSubstitutionEqualsAndHashCodeAreConsistent() {
+        final List<Substitution> substitutions = new ArrayList<>();
+        for (int position : new int[] {0, 1}) {
+            for (byte base : new byte[] {(byte) -1, (byte) 'A'}) {
+                for (byte referenceBase : new byte[] {(byte) -1, (byte) 'C'}) {
+                    for (byte code : new byte[] {Substitution.NO_CODE, (byte) 2}) {
```

Comment:
Add a second valid code here (like 1).  This makes the test fail.

Reply:
This is because two objects with different codes (neither of which is `NO_CODE`) will always compare equal. I think the fix is not to use `code` in the computation of `hashCode`.

Reply:
I think that works.  If the only difference between two of these is that one code is `NO_CODE` but the other isn't, equals() says false but hashCode() is the same.  That's OK in the contract.

Reply:
Right. Technically `hashCode()` could always return the same number and be compliant (that might have saved a lot of this bother!), but then it's not very good for hash tables.